### PR TITLE
Modulefiles generated with a template engine

### PR DESCRIPTION
**Pull request by alalazo**
### TLDR

This PR is a big refactoring effort on module files (with only little changes on the UI side) that introduces the use of template engines in `spack`. Iit relies on `jinja2` that will be shipped as an external, at least until we solve #1385. I also plan to add a lot more tests on hierarchical module files generation.

### Motivation
I started the PR because I needed to extend lmod module file generation to take into account packages that provide more than one service (hello `intel-parallel-studio`!) and figured out that this part of the code has grown enough in size and complexity to benefit from using a template engine for generation.

I also think that this capability may be reused in many places like:

  - @adamjstewart `spack create` command
  - @citibeth `spack loads` and `spack setup` commands
  - generation of JUnit or CDash conformant files from tests (@tgamblin and @kielfriedt may be interested here)

i.e. basically everywhere we need to generate a file with a clear structure starting from the information in a DAG or in the DB.

### Modifications planned (or already done)

**Template engine in spack**
- [x] added an entry `template_dirs` in `config.yaml` to list the directories we should search for templates.
- [x] hooked [jinja2](http://jinja.pocoo.org/docs/2.9/) to spack
- [x] copied [jinja2](http://jinja.pocoo.org/docs/2.9/)  sources into `external`

**Modulefiles**
- [x] added a lot more in-code documentation.
- [x] refactored modulefile generation to use the template engine package (package `spack/modules`).
- [x] added more tests for lmod hierarchies.
- [x] generalized the code (modulo bugs) to multiple providers, i.e. to packages that provide more than one service in the lmod hierarchy.
- [x] implemented the semantic discussed in #3173 (`spack module refresh` permits to generate every module file known, not just the enabled ones).
- [x] implemented an overriding mechanism for the template used to generate module files.
- [x] `spack module refresh` without any specific `--module-type` refreshes all enabled module types.
- [x] added configure line in module files for  `AutotoolsPackage` and `CMakePackage` objects.
- [x] module file short description defaults to the first line of the package docstring (see #1114)
- [ ] restrict the set of tokens of `spec.format` that can be used in the naming scheme (see #2884)
- [x] updated module files tutorial
- [ ] updated reference documentation

**Changes in UI**
- [x] changed keyword `autoload_verbose` to just `verbose` and made it default to `False`. Its location is at the same level as `whitelists` and `blacklists`.

### CODE REVIEW

**Comment by adamjstewart** [comment]
This is awesome! Some things to consider:

1. There are a lot of already open PRs that affect modules. These will have to be rewritten or merged into this PR.

2. Maybe we could just add `jinja2` to `lib/spack/external` until we get better bootstrapping support.

At ANL's LCRC, we are getting ready to make the switch from our own homegrown SoftEnv package to Lmod. I started playing around with Spack's Lmod support and found a few problems worth fixing. I was going to try to take a stab at some of these, but will obviously wait until this PR is done now. You're welcome to try to solve some of these if you like:

1. `spack module refresh` without any specific `--module-type` should refresh all enabled module types. [low priority]

2. Users are always asking me how I built a certain package. It would be neat if we could provide the `configure` or `cmake` line in the module file somehow. [low priority]

3. Right now, when I run `module avail`, I see a giant list of every module I've ever created. NERSC adds some nice formatting that separates things out more. For example, we would want a Core section for things like compilers where we don't care how they were built. We have 2 clusters and 3 MPI fabrics at ANL's LCRC, so we would want to separate those out and possibly only load the MODULEPATH pertaining to that cluster when a user logs in. I'll have to play around a bit more, but I'm not sure if Spack currently handles this case. [medium priority]

4. `spack module refresh` needs to work regardless of whether or not I add a package in a branch and switch back to develop, or if a variant gets added/removed in develop. This is currently the biggest problem I've seen. When a user asks me to install something that isn't in Spack yet, I create a new package and install it. I can't tell them "well, I have a package for it, but I can't install it until the next Spack release, sorry". [high priority]

5. One of the nice features that separates Lmod and Tcl modules is it's ability to control what compilers/MPI libraries are loaded. This involves the use of the [family directive](http://lmod.readthedocs.io/en/latest/010_user.html#users-can-only-load-one-compiler-or-mpi-stack-at-a-time), which prevents you from loading multiple compliers/MPI at the same time. It also allows you to "swap" an entire software stack from one compiler to another with `module swap gcc intel`. I don't think we currently add this family directive. [high priority]

**Comment by alalazo** [comment]
Hey @adamjstewart thanks for the quick feedback! Before I start a by-point reply: feel free to post any problem you encounter with modules (or open an issue an point me to that :smile: ). I'll try to fix what I can here.

##### About this PR

>This is awesome! Some things to consider:
>
> 1. There are a lot of already open PRs that affect modules. These will have to be rewritten or merged into this PR.

I am willing to merge their functionality here (or let the original author of the PR do that if they prefer).  For #2902 this is basically done, as templates are designed to be extended. I still need to check what #2345 does (thanks for the ping yesterday). Others?

> Maybe we could just add jinja2 to lib/spack/external until we get better bootstrapping support.

That would simplify life! I would like to hear what @tgamblin thinks about it. Currently I used only basic features in templates (i.e. the ones that would have permitted to write a replacement for `jinja2` without re-implementing a compiler). If we ship `jinja2` as an external we can remove this limitation.

##### About `lmod` support and issues

> spack module refresh without any specific --module-type should refresh all enabled module types.

I agree, will do

> Users are always asking me how I built a certain package. It would be neat if we could provide the configure or cmake line in the module file somehow

Also agree.

> Right now, when I run module avail, I see a giant list of every module I've ever created. NERSC adds some nice formatting that separates things out more. For example, we would want a Core section for things like compilers where we don't care how they were built. We have 2 clusters and 3 MPI fabrics at ANL's LCRC, so we would want to separate those out and possibly only load the MODULEPATH pertaining to that cluster when a user logs in. I'll have to play around a bit more, but I'm not sure if Spack currently handles this case. [medium priority]

This is currently handled right **for a given architecture**. You need to specify:
```yaml
modules:
  lmod:
    core_compilers:
      - 'gcc@4.8'  # Supposing gcc@4.8 is a core compiler
```
See [the tutorial here](https://spack.readthedocs.io/en/latest/tutorial_sc16_modules.html#lua-hierarchical-module-files).

> spack module refresh needs to work regardless of whether or not I add a package in a branch and switch back to develop, or if a variant gets added/removed in develop. This is currently the biggest problem I've seen. When a user asks me to install something that isn't in Spack yet, I create a new package and install it. I can't tell them "well, I have a package for it, but I can't install it until the next Spack release, sorry"

This problem is orthogonal to what is done here and that's because even if it is revealed by module files, it is inherently a problem of the database returning to you an inconsistent spec-package pair. I think it has to be solved with high priority, but can be done in another PR.

> One of the nice features that separates Lmod and Tcl modules is it's ability to control what compilers/MPI libraries are loaded. This involves the use of the family directive, which prevents you from loading multiple compliers/MPI at the same time. It also allows you to "swap" an entire software stack from one compiler to another with module swap gcc intel. I don't think we currently add this family directive.

Should be already in `develop`. Actually you can put other services into the mix and have a hierarchy like:
```yaml
modules:
  lmod:
    hierarchical_scheme:
      - lapack # Adds lapack on top of MPI/Compiler
```
with the only limitation that there is no support for packages providing two services (in this case both MPI and LAPACK). This will be solved here.

**Comment by adamjstewart** [comment]
Sounds good to me. I did a quick run-through of all open Issues and PRs and labeled any that I thought were related to modules:
https://github.com/LLNL/spack/labels/modules

**Comment by citibeth** [comment]
1. I like StringTemplate because it's the only template engine I've seen that does sub-templates gracefully, rather than dumping you on the floor.  The core observation is that templating is a good fit for functional languages.  I have no idea if StringTemplate would be practial for use in Spack.


2. Simply including jinja2 in Spack (or whatever template engine you decide to use) is goign to be better than building your own.  I agree, we should just include this stuff in Spack for now, and fix the bootstrapping issues two generations later.


> @citibeth `spack module loads` and `spack setup commands`

3. I don't have much interest in this.  I'm happy with how my hacky "Python templates" are working there.  However, if someone else wants to change the way the templating is done, and it doesn't break the current usage, I wouldn't be against it.  I suppose it would make things more flexibility; but so far, I haven't come across a need for that flexibility in the `spconfig.py` file.

4. I don't expect this PR will affect #3134 particularly much, other than the usual needs to rebase / merbe.

**Comment by luigi-calori** [comment]
First of all thanks a lot for the great work on modules.  Using template for module generation was something I had in mind since a while and now here it is ;-)
I' ve had a very quick look at the code, some questions:

1) Would the user be able to override the predefined templates? 

2) In our current custom modules deployment,  some modules require specific code to be added.
would it be possible to define a template that expand a custom field that is declared in some package.py file. Otherwise,  the custom code should be specified in an include file defined in module.yaml

3) On our cluster we have old, pure tcl module setup. In order to ease spack adoption:
   Our module system do not have modulecmd, so we can not use modules loading in package.yaml    
   we addressed  #2924  by resurrecting and modifying #2426 . I have an WIP PR for that:
   Ok to submit?
   We currently use some module grouping ( profile ) to reduce the number of entries presented by
   `module av` . Is there a plan to support module path  grouping based on some custom info,
   either added in modules.yaml or in individual package.py

**Comment by citibeth** [comment]
Also, we would need to get #2664 merged before this PR messes with `spack setup` templates.

**Comment by alalazo** [comment]
@citibeth I don't plan to touch commands here.

**Comment by alalazo** [comment]
@luigi-calori 

> Would the user be able to override the predefined templates?

The mechanism currently implemented permits you to specify a list of `template_dirs` in the `config.yaml` file like:
```yaml
config:
  # Locations where templates should be found
  template_dirs:
    - $spack/templates
```
To override templates for module files you have then two possibilities. 

The first one is at package level, if it makes sense to do so. In this case you need to set e.g. the `tcl_template` attribute to override the template for `tcl`. 

The second possibility (with higher precedence) is to specify the template in the `modules.yaml` file under an anonymous spec entry. See the unit tests for more insight on this.

> In our current custom modules deployment, some modules require specific code to be added.
would it be possible to define a template that expand a custom field that is declared in some package.py file. Otherwise, the custom code should be specified in an include file defined in module.yaml

You can use all the powers of [jinja2](http://jinja.pocoo.org/docs/2.9/templates/) and the mechanisms above to define custom code for your templates. 

> On our cluster we have old, pure tcl module setup. In order to ease spack adoption:
Our module system do not have modulecmd, so we can not use modules loading in package.yaml
we addressed #2924 by resurrecting and modifying #2426 . I have an WIP PR for that:
Ok to submit?

It's **always ok** to submit a PR. Maybe you want to ask people in #2924 for an update on the state of things. Do you think that work will be related to this refactoring?

>We currently use some module grouping ( profile ) to reduce the number of entries presented by
module av . Is there a plan to support module path grouping based on some custom info,
either added in modules.yaml or in individual package.py

Start using lmod! :smile:

**Comment by alalazo** [comment]
@TheQueasle I think you may be interested in this PR as it should extend the functionality introduced in #2902 . Can you tell me if it fits your use case?

**Comment by TheQueasle** [comment]
@alalazo Cursory glance things look good.  I'll need to carve out some time to play around with this to be 100% sure.

Honestly though, so long as I can inject arbitrary code in the generated modulefiles it will suit me just fine, and it looks like you covered that usecase already. :smile:

**Comment by alalazo** [comment]
@adamjstewart The latest commit should address the change in behavior for `spack module refresh` when called without specifying the module type. I also changed `spack module rm` for consistency.

File: lib/spack/spack/cmd/common/arguments.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def _specs(self, **kwargs):
 _arguments['module_type'] = Args(
     '-m', '--module-type',
     choices=spack.modules.module_types.keys(),
-    default=spack.modules.module_types.keys()[0],
+    action='append',
     help='type of module files [default: %(default)s]')
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/common/arguments.py]: What's your plan with the default? Also, if you change the default, make sure to change the help message.

Comment (alalazo) [inline lib/spack/spack/cmd/common/arguments.py]: Good catch on the help message. I changed the action to `append` so that one can do:
```
spack module refresh -m tcl -m lmod ...
```
The default is the one you proposed (all the module files that are known) except that it's set in the method `module`.

**Comment by hartzell** [comment]
I thought I'd take this for a test drive and see if I "liked" the modulefiles it generated.  I clone the *epfl-scitas/spack* repo, checked out the *refactoring/modulefiles_genreated_with_a_template_engine* branch and tried to run spack.

This is the result (CentoOS 7).

```
[hartzelg@lb117esx spack]$ ./bin/spack
Traceback (most recent call last):
  File "./bin/spack", line 137, in <module>
    module = spack.cmd.get_module(cmd)
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/cmd/__init__.py", line 87, in get_module
    level=0)
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/cmd/build.py", line 25, in <module>
    import spack.cmd.configure as cfg
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/cmd/configure.py", line 29, in <module>
    import spack.cmd.install as inst
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/cmd/install.py", line 38, in <module>
    import spack.cmd.common.arguments as arguments
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/cmd/common/arguments.py", line 30, in <module>
    import spack.modules
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/modules/__init__.py", line 32, in <module>
    from .dotkit import DotkitModulefileWriter
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/modules/dotkit.py", line 28, in <module>
    from .common import BaseConfiguration, BaseFileLayout
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/modules/common.py", line 62, in <module>
    import spack.tengine as tengine
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/spack/tengine.py", line 28, in <module>
    import jinja2
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/external/jinja2/__init__.py", line 33, in <module>
    from jinja2.environment import Environment, Template
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/external/jinja2/environment.py", line 15, in <module>
    from jinja2 import nodes
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/external/jinja2/nodes.py", line 19, in <module>
    from jinja2.utils import Markup
  File "/home/hartzelg/tmp/alalazo/spack/lib/spack/external/jinja2/utils.py", line 624, in <module>
    from markupsafe import Markup, escape, soft_unicode
ImportError: No module named markupsafe
[hartzelg@lb117esx spack]$
```

I'm probably missing something obvious....

**Comment by alalazo** [comment]
@hartzell Thank you. It seems I didn't add some externals that are probably a dependency of `jinja2`. I also have to re-check the tutorial on modulefiles and modify slightly the docs. If you agree with it I'll ping you when done. Thanks for volunteering!

**Comment by alalazo** [comment]
Note to others: oddly enough, if the same branch is used in two separate pull requests, Travis is triggered randomly on only one of the two...

**Comment by hartzell** [comment]
@alalazo -- happy to give it a review.

**Comment by alalazo** [comment]
@hartzell Perfect timing: I think the PR is ready for a first round. 

I still need to update the reference docs, but the tutorial should be fine. If you have time I am interested in receiving feedback on the readability of the code under `lib/spack/spack/modules/*` and of the templates under `templates`. 

Hierarchical module files are a bit slow to regenerate the entire tree after the generalization I have done. I may also look into optimizing that if I can.

**Comment by hartzell** [comment]
It seems to generate reasonable modules for everything that's in my standard tree.  Quick testing hasn't revealed any issues.

I'll take a look at the docs and code.

It takes 45 seconds to regenerate my tree, which I'll do very rarely.  Optimizing that doesn't seem like a high priority.

**Comment by alalazo** [comment]
@hartzell Are you using `tcl` or `lmod`? What I generalized is just `lmod` that now can account for packages that provide more than one service (mpi, lapack, etc...) in the hierarchy. `tcl` shouldn't take that long to regenerate, `lmod` takes sensibly longer.

**Comment by hartzell** [comment]
I'm using `lmod`.

File: templates/modules/modulefile.lua [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+{% elif command_name == 'SetEnv' %}
+setenv("{{ cmd.name }}", "{{ cmd.value }}")
+{% elif command_name == 'UnsetEnv' %}
+unsetenv("{{ cmd.name }}")
+{% endif %}
+{% endfor %}
+{% endblock %}
+
+{% block footer %}
+{# In case he module needs to be extended with custom LUA code #}
```
Comment (hartzell) [inline templates/modules/modulefile.lua]: "In case **the** module..."

File: templates/modules/modulefile.lua [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+if not isloaded("{{ module }}") then
+{% if verbose %}
+    LmodMessage("Autoloading {{ module }}")
+{% endif %}
+    load("{{ module }}")
+end
+{% endfor %}
+{% endblock %}
+
+{# {% block prerequisite %}
```
Comment (hartzell) [inline templates/modules/modulefile.lua]: Is it worth commenting on why this is commented out?  Perhaps comment it out more blatently?

File: lib/spack/spack/modules/lmod.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+#: LMOD specific part of the configuration
+configuration = configuration.get('lmod', {})
+
+#: Caches the configuration {spec_hash: configuration}
+configuration_registry = {}
+
+
+def make_configuration(spec):
+    """Returns the tcl configuration for spec"""
```
Comment (hartzell) [inline lib/spack/spack/modules/lmod.py]: "Returns the lmod configuration...",

File: lib/spack/spack/modules/common.py [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+
+and is divided among four classes:
+
+  * a configuration class that provides a convenient interface to query
+    configuration for the spec under consideration.
+
+  * a layout class that provides the information associated with module
+    file names and directories
+
+  * a context class that provides the dictionary used byt the template engine
```
Comment (hartzell) [inline lib/spack/spack/modules/common.py]: "used by the template..."

**Comment by hartzell** [comment]
@alalazo -- I fixed a couple of trivial typos as I read through it.  It makes sense to me and I believe that if I had a particular need in mind I probably get there from the code and the comments.

File: templates/modules/modulefile.lua (hunk 3 above)
Comment (alalazo) [inline templates/modules/modulefile.lua]: I won't tell that it is commented because I thought about implementing the feature and then forgot about it... no I won't tell anybody. :smile:

**Comment by alalazo** [comment]
Thanks for the feedbck @hartzell. I'll try to write up what's missing in the docs and implement `prereq` as soon as I can.

File: templates/modules/modulefile.lua (hunk 3 above)
Comment (hartzell) [inline templates/modules/modulefile.lua]: 🙊

**Comment by alalazo** [comment]
@adamjstewart I was wondering: do you mind reviewing this PR? 

It's basically complete from my point of view (except the `prereq` in `lmod`, see above). I updated the existing docs. It just lacks new documentation on how to to use the template system if you want to extend or substitute the template module file for a particular set of specs.

**Comment by adamjstewart** [comment]
I can try to get to it by the end of the week.

File: lib/spack/spack/cmd/module.py [hunk 6]
```diff
@@ -27,24 +27,23 @@
...
-# Each sub-command must be callable and accept 3 arguments :
-# - mtype : the type of the module file
-# - specs : the list of specs to be processed
-# - args : namespace containing the parsed command line arguments
+#: Dictionary that will be populated with the list of sub-commands
+#: Each sub-command must be callable and accept 3 arguments:
+#:
+#:   - mtype : the type of the module file
+#:   - specs : the list of specs to be processed
+#:   - args : namespace containing the parsed command line arguments
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/module.py]: I've been meaning to ask this, but why do you put colons after your comment symbols?

File: lib/spack/spack/cmd/module.py [hunk 7]
```diff
@@ -102,17 +101,54 @@ def setup_parser(subparser):
     )
 
 
-class MultipleMatches(Exception):
-    pass
+class MultipleSpecsMatch(Exception):
+    """Raised when multiple specs match a constraint, in a context where
+    this is not allowed.
+    """
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/module.py]: Don't these still need `pass`?

File: lib/spack/spack/cmd/module.py [hunk 8]
```diff
@@ -102,17 +101,54 @@ def setup_parser(subparser):
...
 
+def one_spec_or_raise(specs):
+    """Ensures exactly one spec has been selected, or raises the appropriate
+    exception.
+    """
+    # Ensure a single spec matches the constraint
+    if len(specs) == 0:
+        raise NoSpecMatches()
+    if len(specs) > 1:
+        raise MultipleSpecsMatch()
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/module.py]: So many people are trying to solve the same thing right now 😆

File: lib/spack/spack/cmd/module.py [hunk 9]
```diff
@@ -155,102 +193,118 @@ def loads(mtype, specs, args):
...
-    if not args.yes_to_all:
-        tty.msg(
-            'You are about to remove {0} module files the following specs:\n'
-            .format(mtype))
-        spack.cmd.display_specs(specs_with_modules, long=True)
-        print('')
-        answer = tty.get_yes_or_no('Do you want to proceed?')
-        if not answer:
-            tty.die('Will not remove any module files')
+def rm(module_types, specs, args):
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/module.py]: Let's call it `remove` or `delete`.

File: lib/spack/spack/environment.py [hunk 10]
```diff
@@ -449,17 +449,36 @@ def validate(env, errstream):
...
 
-    Args:
-        env: list of environment modifications
-        variables: list of variable names to be filtered
+    :param env: list of environment modifications
+    :param variables: list of variable names to be filtered
 
-    Yields:
-        items in env if they are not in variables
+    :return: items in env if they are not in variables
```
Comment (adamjstewart) [inline lib/spack/spack/environment.py]: These can be converted back to Google-style docstrings now.

File: lib/spack/spack/modules/common.py [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+    return [x for x in l if not (x in seen or seen_add(x))]
+
+
+def merge_config_rules(configuration, spec):
+    """Parses the module specific part of a configuration and returns a
+    dictionary containing the actions to be performed on the spec passed as
+    an argument.
+
+    :param configuration: module specific configuration (e.g. entries under
+        the top-level 'tcl' key
```
Comment (adamjstewart) [inline lib/spack/spack/modules/common.py]: Missing an end parenthesis

File: lib/spack/spack/modules/dotkit.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""This module implements the classes necessary to generate dotkit modules."""
+import os.path
+
+from .common import BaseConfiguration, BaseFileLayout
+from .common import BaseContext, BaseModuleFileWriter, configuration
+
+#: Dotkit specific part of the configuration
+configuration = configuration.get('dotkit', {})
```
Comment (adamjstewart) [inline lib/spack/spack/modules/dotkit.py]: I'm not sure how many people use multiple module systems, but it would be cool if I could do something like:
```yaml
modules:
  enable:
    - lmod
    - tcl
    - dotkit
  all:
    # settings for all module systems
  tcl:
    # tcl-specific settings
```
This would be in line with how `packages.yaml` works.

File: lib/spack/spack/modules/lmod.py [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+    def core_compilers(self):
+        """Returns the list of "Core" compilers
+
+        :raises CoreCompilersNotFoundError: if the key was not
+            specified in the configuration file or the sequence
+            is empty
+        """
+        value = configuration.get('core_compilers')
+        if value is None:
+            msg = '\'core_compilers\' key not found in configuration file'
```
Comment (adamjstewart) [inline lib/spack/spack/modules/lmod.py]: Just put single quotes inside of double quotes, backlashes are ugly.

File: lib/spack/spack/modules/lmod.py [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+            raise CoreCompilersNotFoundError(msg)
+        if not value:
+            msg = '\'core_compilers\' list cannot be empty'
+            raise CoreCompilersNotFoundError(msg)
+        return value
+
+    @property
+    def hierarchy_tokens(self):
+        """Returns the list of tokens that are part of the modulefile
+        hierarchy. 'compiler' is always present.
```
Comment (adamjstewart) [inline lib/spack/spack/modules/lmod.py]: What would happen if `compiler` was not always present? I know it sounds silly, as the whole point of Lmod is that it's hierarchical. Just thinking about handling weird use cases.

File: lib/spack/spack/tengine.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
+##############################################################################
```
Comment (adamjstewart) [inline lib/spack/spack/tengine.py]: I would call the file `template_engine.py`.

File: .codecov.yml [hunk 16]
```diff
@@ -13,6 +13,10 @@ coverage:
         threshold: 0.5
         paths:
           - lib/spack/spack/cmd
+      modules:
+        threshold: 0.5
+        paths:
+          - lib/spack/spack/modules
```
Comment (alalazo) [inline .codecov.yml]: This gives clear indication, imho, of the coverage of a spack package (in the python sense, so folder with `__init__.py` to be rough :smile: ). Modules are covered 91% apparently,  and if I click on the summary I'll see immediately where I lack testing.

Comment (adamjstewart) [inline .codecov.yml]: That is pretty cool. Maybe we should add this for every directory in there.

File: lib/spack/docs/module_file_support.rst [hunk 17]
```diff
@@ -296,65 +271,68 @@ For example, consider the following on one system:
...
 
-Though Spack ships with sensible defaults for the generation of module files,
-one can customize many aspects of it to accommodate package or site specific needs.
-These customizations are enabled by either:
+Spack ships with sensible defaults for the generation of module files, but
+you can customize many aspects of it to accommodate package or site specific needs.
+In general you can override or extend the default behavior by:
 
  1. overriding certain callback APIs in the Python packages
  2. writing specific rules in the ``modules.yaml`` configuration file
```
Comment (alalazo) [inline lib/spack/docs/module_file_support.rst]: Ehehe, sometimes you have 15.000 lines in front of you an you don't see them....

File: lib/spack/spack/cmd/module.py (hunk 6 above)
Comment (alalazo) [inline lib/spack/spack/cmd/module.py]: Because it should appear in the docs for `callbacks` below. Even though honestly I didn't check this particular variable.

File: lib/spack/spack/cmd/module.py (hunk 7 above)
Comment (alalazo) [inline lib/spack/spack/cmd/module.py]: I think just a `docstring` is fine either. We should have errors coming from the interpreter otherwise.

File: lib/spack/spack/modules/lmod.py (hunk 14 above)
Comment (alalazo) [inline lib/spack/spack/modules/lmod.py]: Maybe I worded the sentence in an unclear way. What I mean is that at least you get back the compiler part. Because that can be inferred from the spec and **needs** to be part of the hierarchy.

**Comment by alalazo** [comment]
@adamjstewart Thanks for the initial review, I'll get to your comments asap.

File: lib/spack/spack/cmd/module.py (hunk 6 above)
Comment (adamjstewart) [inline lib/spack/spack/cmd/module.py]: OOoooooh!!! I was wondering how to do this! Can you point me to the Sphinx docs on this feature?

Comment (alalazo) [inline lib/spack/spack/cmd/module.py]: Serach for `#:` [here](http://www.sphinx-doc.org/en/stable/ext/autodoc.html#module-sphinx.ext.autodoc)

**Comment by adamjstewart** [comment]
Another feature that I've been thinking about for a long time that this could be used for is a way to deprecate certain modules. Basically, when you load this module, it should print a warning message like:
```
/intel/17.0.3/mvapich2/2.2b-fdskgs is deprecated, use /intel/17.0.3/mvapich2/2.2-kajsdf instead.
```
I could hack the module files myself to do this, but it would be blown away every time I refreshed (not sure how often I'll need to do this though). I can't decide the best way to do this. Has anyone else ever done this? What's the recommended approach?

**Comment by alalazo** [comment]
@adamjstewart Probably we'll need to read [this](http://lmod.readthedocs.io/en/latest/140_deprecating_modules.html) to decide how to deal with it in Spack.

**Comment by adamjstewart** [comment]
They really have thought of everything, haven't they.

File: lib/spack/spack/modules/dotkit.py (hunk 12 above)
Comment (alalazo) [inline lib/spack/spack/modules/dotkit.py]: I think I'll keep this one for a subsequent PR.

File: lib/spack/spack/tengine.py (hunk 15 above)
Comment (alalazo) [inline lib/spack/spack/tengine.py]: I can't decide on this. On the one hand I agree that the name is not really clear. On the other hand I really dislike module names that are more than one word. Anybody with a suggestion that is both short and clear?

File: lib/spack/docs/module_file_support.rst [hunk 18]
```diff
@@ -574,11 +549,14 @@ most likely via the ``+blas`` variant specification.
...
-     that will generate a hierarchy in which the ``lapack`` layer is treated as the ``mpi``
-     one. This allows a site to build the same libraries or applications against different
+           core_compilers:
+             - 'gcc@4.8'
+           hierarchy:
+             - 'mpi'
+             - 'lapack'
+
+     that will generate a hierarchy in which the ``lapack`` and ``mpi`` layer can be switched
+     independently. This allows a site to build the same libraries or applications against different
```
Comment (alalazo) [inline lib/spack/docs/module_file_support.rst]: I added:
```yaml
modules:
  lmod:
    hierarchy:
      - mpi
```
to ``modules.yaml``.

File: templates/modules/modulefile.lua (hunk 3 above)
Comment (alalazo) [inline templates/modules/modulefile.lua]: @hartzell I think this PR is already big enough, so I just deleted the comments.  I plan to implement the feature in a smaller PR later on (after this is merged).

**Comment by adamjstewart** [comment]
@alalazo Quick question. I think I've seen a couple open PRs for this, but how do I rename a module file like:
```
intel/17.0.3-gbgtimn
```
to
```
intel/17.0.3/gbgtimn
```
Since Spack's hashing mechanism changes so frequently, I want users to be able to load `intel/17.0.3` and not worry about the exact hash. This is inline with [Lmod's NVV hierarchy](http://lmod.readthedocs.io/en/latest/060_locating.html#nvv-picking-modules-when-there-are-multiple-directories-in-modulepath).

**Comment by alalazo** [comment]
@becker33 @scheibelp Any chance you have time to review this PR, so that we can push towards merging it?

In the incoming weeks I'll have some more time to devote to development and would like to improve a bit module file generation. It would be nice if I could clean this PR and get it accepted, and then build on top of it.

**Comment by alalazo** [comment]
@adamjstewart I think you may want to check the commit 22d61df. It solves a bug in 7cec91a that was revealed by building `r` packages. If you don't have major concerns over the interface I prefer the solution above to either skipping this information or treating `RPackage` and `WafPackage` differently from `AutotoolsPackage`.

Sidenote: that modification also applies to `PythonPackage` but as it is not directly involved with the bug I restrained from modifying that part here.

**Comment by adamjstewart** [comment]
@alalazo Can you give me more specifics on the bug?

**Comment by alalazo** [comment]
@adamjstewart The bug in the module file PR is that I try to get the configure args from packages to put them in the modules, and call the method with no args:
```python
    @tengine.context_property
    def configure_options(self):
        pkg = self.spec.package

        # This is quite simple right now, but contains information on how
        # to call different build system classes.
        for attr in ('configure_args', 'cmake_args'):
            try:
                configure_args = getattr(pkg, attr)()
                return ' '.join(configure_args)
            except (AttributeError, IOError):
                pass

        # The default is to return None
        return None
```
I could except also on `TypeError`, but that means we get no config options for R and Waf packages. Or I could inspect the configure signature to see how many args it takes, but that complicates a lot the logic above for a small gain.

**Comment by adamjstewart** [comment]
Okay, so there's nothing wrong with using `(self)` or `(self, spec, prefix)`, just mixing both of them. I still argue for the latter, but I understand your reasons for the former.
